### DISCUSSION TITLE
default theme 404 error message customization

### DISCUSSION THREAD

**Discussion by tassoman** [vuejs/vitepress#2169] [Q&A] [answered]
It's possible to customize the 404 error message, without writing a custom theme?

For example by having just a configuration, also compatible with i18n

```typescript
themeConfig: {
  404: {
    title: 'Page not found',
    message: [
      "But if you don't change your direction, and if you keep looking, you may end up where you are heading.",
      "Another random message",
    ],
    action: 'Take me home',
    link: '/',
  },
},
```

**Comment by brc-dd** [comment]
`not-found` slot? https://vitepress.dev/guide/extending-default-theme#layout-slots

**Reply by tassoman** [reply]
I was inspired by this [Vuepress configuration](https://v2.vuepress.vuejs.org/reference/default-theme/config.html#notfound).

Extending `not-found` slot means to write a component into a custom theme, I can understand well?

**Reply by brc-dd** [reply]
Yeah, you'll just have to write a component and add 3-4 lines to extend the default theme and add that component to slot, and then you'll get it on the 404 page.

**Reply by tassoman** [reply]
I managed to write the Layout extender, who reads mini-templates from MyLayout extension.

when I edit the `not-found` template, do I overwrite this component? [NotFound.vue](https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/NotFound.vue)

**Reply by brc-dd** [reply]
> when I edit the not-found template, do I overwrite this component? [NotFound.vue](https://github.com/vuejs/vitepress/blob/main/src/client/theme-default/NotFound.vue)

yeah

**Reply by tassoman** [reply]
While for the hooks `*-before` and `*-after` the extension is pretty straightforward, overriding the whole template it's overwhelmed complicated because needs to bring the whole parts of a Vue component. Maybe the [component override procedure](https://vitepress.dev/guide/extending-default-theme#overriding-internal-components) is more suitable. 

Now I have overwritten but I can't figure how I could i18n the whole thing ... I just expected something easier or pluggable like a configuration.

**Comment by tassoman** [comment]
@brc-dd I'm sorry I return on this question.

Is there a way to simply override `{{ theme.notFound.* }}` variables, as seen here 👀   
Maybe in `themeConfig` ? 

https://github.com/vuejs/vitepress/blob/7a3b2d591bf187ae7f444d809bb5c96aeb7178d9/src/client/theme-default/NotFound.vue#L25

**Reply by brc-dd** [reply, marked as answer]
Yeah, those weren't added at that time. On rc versions you do that:

```ts
// .vitepress/config.ts

export default defineConfig({
  themeConfig: {
    notFound: {
      title: 'PAGE NOT FOUND',
      quote: "But if you don't change your direction, and if you keep looking, you may end up where you are heading.", // set to '' to hide
      linkLabel: 'go to home', // aria-label
      linkText: 'Take me home',
      code: '404'
    }
  }
})
```

You can customize them per locale too if you're using i18n:


```ts
import { defineConfig } from 'vitepress'

export default defineConfig({
  // shared properties and other top-level stuff...

  locales: {
    root: {
      label: 'English',
      lang: 'en',

      themeConfig: { notFound: { ... } }
    },
    fr: {
      label: 'French',
      lang: 'fr',

      themeConfig: { notFound: { ... } }
    }
  }
})
```

**Reply by tassoman** [reply]
cool, thanks!
